File: modules/admin/admin-ui/src/main/resources/web/admin/common/js/issue/PublishRequest.ts
module api.issue {

    import PublishRequestJson = api.issue.resource.PublishRequestJson;

    export class PublishRequest {

        private excludeIds: ContentId[];

        private items: PublishRequestItem[];

        constructor(builder: PublishRequestBuilder) {
            this.excludeIds = builder.excludeIds;
            this.items = builder.issueItems;
        }

        public getExcludeIds(): ContentId[] {
            return this.excludeIds;
        }

        public getItems(): PublishRequestItem[] {
            return this.items;
        }

        public getItemsIds(): ContentId[] {
            return this.items ?
                   this.items.map(item => item.getId()) : [];
        }

        public getExcludeChildrenIds(): ContentId[] {
            return this.items ?
                   this.items.filter(item => !item.isIncludeChildren()).map(item => item.getId()) : [];
        }

        toJson(): PublishRequestJson {
            return {
                excludeIds: this.excludeIds.map(id => id.toString()),
                items: this.items.map(id => id.toJson()),
            };
        }

        public static create(): PublishRequestBuilder {
            return new PublishRequestBuilder();
        }

    }

    export class PublishRequestBuilder {

        excludeIds: ContentId[] = [];

        issueItems: PublishRequestItem[] = [];

        fromJson(json: PublishRequestJson): PublishRequestBuilder {
            this.excludeIds = json.excludeIds ? json.excludeIds.map(excludeId => new ContentId(excludeId)) : [];
            this.issueItems = json.items ?
                              json.items.map(itemJson => PublishRequestItem.create().fromJson(itemJson).build()) : [];

            return this;
        }

        public addExcludeId(id: ContentId): PublishRequestBuilder {
            this.excludeIds.push(id);
            return this;
        }

        public addExcludeIds(ids: ContentId[] = []): PublishRequestBuilder {
            this.excludeIds = this.excludeIds.concat(ids);
            return this;
        }

        public addPublishRequestItem(item: PublishRequestItem): PublishRequestBuilder {
            this.issueItems.push(item);
            return this;
        }

        public addPublishRequestItems(items: PublishRequestItem[] = []): PublishRequestBuilder {
            this.issueItems = this.issueItems.concat(items);
            return this;
        }

        public build(): PublishRequest {
            return new PublishRequest(this);
        }
    }
}

File: modules/core/core-api/src/main/java/com/enonic/xp/issue/PublishRequest.java
package com.enonic.xp.issue;

import com.enonic.xp.content.ContentId;
import com.enonic.xp.content.ContentIds;

public final class PublishRequest
{
    private final ContentIds excludeIds;

    private final PublishRequestItems items;

    private PublishRequest( Builder builder )
    {
        this.items = builder.items.build();
        this.excludeIds = builder.exclude.build();
    }

    public PublishRequestItems getItems()
    {
        return items;
    }

    public ContentIds getExcludeIds()
    {
        return excludeIds;
    }

    public static Builder create()
    {
        return new Builder();
    }

    public static final class Builder
    {
        private final PublishRequestItems.Builder items = PublishRequestItems.create();

        private final ContentIds.Builder exclude = ContentIds.create();

        private Builder()
        {
        }

        public Builder addItems( final PublishRequestItems items )
        {
            this.items.addAll( items );
            return this;
        }

        public Builder addItem( final PublishRequestItem item )
        {
            this.items.add( item );
            return this;
        }

        public Builder addExcludeIds( final ContentIds exclude )
        {
            this.exclude.addAll( exclude );
            return this;
        }

        public Builder addExcludeId( final ContentId exclude )
        {
            this.exclude.add( exclude );
            return this;
        }

        public PublishRequest build()
        {
            return new PublishRequest( this );
        }
    }
}
